src/com/amazon/corretto/crypto/provider/EvpKey.java

```diff
@@ -216,24 +222,81 @@ protected interface CanDerivePublicKey<T extends EvpKey & PublicKey> {
+   *
+   * Instead, we use writeReplace() and readResolve() to store our information in a dedicated format for serialization.
+   * - writeReplace() on the object to be serialized (an EvpKey) returns the *real* object to be serialized.
+   *   In our case, we return an instance of SerializedKey which contains the minimal information to properly save and retrieve a key.
+   * - readResolve() is called on the object which is *actually* serialized (and thus being deserialized) and returns the object that users care about.
+   *   In our case, SeralizedKey.readResolve() returns an appropriate instance of EvpKey.
+   *
+   * So, the entire (hidden from the user) flow goes like this.
+   * 1.  User tries to serialize an EvpKey
+   * 2.  Java detects that EvpKey.writeReplace() exists and calls it
+   * 3.  EvpKey.writeReplace() returns an instance of SerializedKey
+   * 4.  Java actually serializes SerializedKey
+   * 5.  (Some time passes)
+   * 6.  User tries to deserialize the bytes
+   * 7.  Java detects that the bytes contain a serialized SerializedKey and so deserializes it.
+   * 8.  Java detects that SerializedKey.readResolve() exists and calls it
+   * 9.  SerializedKey.readResolve() creates and returns an instance of EvpKey
+   * 10. User gets an instance of EvpKey and never realizes that the above complexity exists
+   */
+  Object writeReplace() throws ObjectStreamException {
+    return new SerializedKey(type, isPublicKey, internalGetEncoded());
+  }
+
+  // This object must never be serialized directly
   private void writeObject(final ObjectOutputStream out) throws IOException {
     throw new NotSerializableException("EvpKey");
   }
 
+  // This object must never be deserialized directly
   private void readObject(final ObjectInputStream in) throws IOException, ClassNotFoundException {
     throw new NotSerializableException("EvpKey");
   }
 
+  // This object must never be deserialized directly
   private void readObjectNoData() throws ObjectStreamException {
     throw new NotSerializableException("EvpKey");
   }
+
+  /**
+   * Minimal information needed to serialize/deserialize any instance of EvpKey. Contains the key
+   * type, whether it is a public key, and the appropriate encoding of it.
+   */
+  private static class SerializedKey implements Serializable {
+    private static final long serialVersionUID = 1;
+    private final EvpKeyType type;
+    private final boolean isPublicKey;
+    private final byte[] encoded;
+
+    public SerializedKey(EvpKeyType type, boolean isPublicKey, byte[] encoded) {
+      this.type = type;
+      this.isPublicKey = isPublicKey;
+      this.encoded = encoded;
+    }
+
+    private Object readResolve() throws ObjectStreamException {
+      try {
+        final KeyFactory kf = AmazonCorrettoCryptoProvider.INSTANCE.getKeyFactory(type);
+        final EvpKey result;
+        if (isPublicKey) {
```

Comment:
Is the type cast needed?

Reply:
Needed? No. Present to make the return type slightly clearer (in that it *must* assignable to an `EvpKey`.) I'll change it to `Key` though which is the next best type.

Reply:
or just `return isPublicKey ? kf.generatePublic(new X509EncodedKeySpec(encoded)) : kf.generatePrivate(new PKCS8EncodedKeySpec(encoded));`